### ISSUE TITLE
Opening CPU sample in Perfview show multiple "processes"  for the same PID

### ISSUE THREAD

**Issue by gleocadie** [microsoft/perfview#1254]
We are evaluating the .Net runtime 5 preview7, and I see a change in behavior when opening cpu samples (perf.data.txt file inside a *.trace.zip archive) in Perfview.

I collected cpu samples (using perf tools) of an application running on Linux with the runtime 5 preview 7. I noticed that in the perf.data.txt file the first line does not contain only the process name, but the name of the thread instead (when the thread has a name, otherwise the process name).

```
.NET ThreadPool    84/2671  72906.063991:    1001001
   0x534535
   ...
dotnet    84/155   72906.370060:    1001001
   0x2422
   ...
Ghost.ThreadPoo    84/498   72908.177015:    1001001
   0x654645
   ...
.NET Server GC    84/702   72908.227497:    1001001
   0x432411
   ...
```
As you can see the process Id (84) is the same in every "header" line, but the corresponding name is different.

I opened it in Perfview (putting it in a `trace.zip` file, see attached) and I got this:
![image](https://user-images.githubusercontent.com/17292193/91635874-482a8d00-e9fc-11ea-9c6f-e54b2538a880.png)

When collecting cpu samples from a 3.1 runtime, it's always `dotnet` for the name. It seems something has changed (maybe a bug fix) in the runtime since 3.X, but it feels like that something should handled in Perfview.
I was expecting in Perfview to display one line per process and not one line per named-thread. Am I missing something? is it the expected behavior ? 

[trace_sample.trace.zip](https://github.com/microsoft/perfview/files/5145127/trace_sample.trace.zip)

**Comment by brianrob** [comment]
Yup @gleocadie, this is expected.  In .NET 5 many of the standard threads have been named.  There is similar functionality on Windows, but it is surfaced in the ETW trace differently, which is why it doesn't show up in the processes dialog.

On Linux, the process name that perf displays gets hijacked by the "thread name".  We can definitely do something here, but the real question is what.  One thing that we could consider is using the process name from the very first sample that we encounter.  This would likely work most of the time for traces that capture process start, but would be hit or miss for those that capture in the middle of the process, where the first sample is very likely to be on a "named" thread.

Another potential option would be to merge the names together and pull out the well known ones that we know aren't the process names (e.g. .NET ThreadPool, .NET Timer, etc.)  This wouldn't be perfect, but would definitely get us closer.  Is there a specific scheme that you have in mind?

**Comment by gleocadie** [comment]
Hello @brianrob , sorry for the late reply.

I was thinking that the perfcollect script could generate a file in which we associate a PID to a process name (from the perf.data.txt file), and save this file in the .trace.zip artifact. 

When Perfview opens the .trace.zip file, we could take this into a account. If this file is not present, we could go for your second potential option (merge the names).

With this file, we could even further: on linux process name is always "dotnet", but if you have multiple .NET process, it becomes hard to differentiate them when opening in Perfview. We could extract the dll file used in the command line and as the process name.

What do you think ?

**Comment by brianrob** [comment]
Sure, I'm all for capturing a bit more context information.  I don't think that the process name is in perf.data.txt in a different spot than where we get it today for display in PerfView.  So, I don't think you'll be able to get what you want from perf.data.txt.  I am wondering if there's a good way to get the command line for each process in the trace - that would be an awesome disambiguator.  We could implement a step at the beginning and/or end of the trace to get these, but I'm not sure of a nice way to get them for traces that get spawned during the trace - there probably is a reasonable way to do this, but not sure.  What do you think?

**Comment by gleocadie** [comment]
To retrieve the name of the process and the command-line, I was thinking about  _**/proc/<pid>/cmdline**_. (we could also look at **_/proc/<pid>/status_** which contains the process name).

Here what I was thinking, let me know what you think:
- Once the trace is done, get all the PIDs from the perf.data.txt
- For all the PIDs, read from _**/proc/<pid>/cmdline**_ and/or **_/proc/<pid>/status_** to get the process name and the command-line.  
- Generate a file which associates the pid to this infos
- Add this file to the archive.

For processes that get spawned during the trace or died during the trace, we could just for the "merge names" option at first.

In Perfview, we will rely on this file and the perf.data.txt (in case we need to merge the names) to have a meaningful name for a given PID.

What do you think ?

PS: I found [execsnoop](https://raw.githubusercontent.com/brendangregg/perf-tools/master/execsnoop) , but the script needs to be run as root to be effective. That could be an option.

**Comment by brianrob** [comment]
@gleocadie, thanks for this.  I think that doing what you suggest seems reasonable.  As you point out, it's not perfect, but it should handle many cases, and I think that's worth trying as a first attempt.  We can refine from there.  execsnoop is also interesting, and running as root is OK - we force that right now, but I'd like to avoid adding another dependency we have to package, so I think skipping that for now is a good choice.
